### PR TITLE
Migrate to near-sdk 4.0 and fix builds

### PR DESCRIPTION
**Pull request by karim-en**
Update `near-sdk` to 4.0

### CODE REVIEW

**Review by sept-en — CHANGES_REQUESTED** [review]

File: bridge-common/Cargo.toml [hunk 1]
```diff
@@ -7,12 +7,11 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-eth-types =  { git = "https://github.com/near/rainbow-bridge", rev = "4abc8d9b28d4a9a62b90d35147218a9d08f6721b" }
+eth-types =  { git = "https://github.com/near/rainbow-bridge", rev = "8644d02031250470556b07fdf3506f13357eb8b7" }
+admin-controlled = { git = "https://github.com/near/rainbow-bridge", rev = "8644d02031250470556b07fdf3506f13357eb8b7" }
```
[the remark this answers is not in the document]
Comment (karim-en) [reply bridge-common/Cargo.toml]: Fixed 29c09b8654d6152708a3608d6b39a4e0e7dcdf1e

File: bridge-token/src/lib.rs [hunk 2]
```diff
@@ -4,19 +4,14 @@ use near_contract_standards::fungible_token::metadata::{
...
 };
-use std::convert::TryInto;
-
-near_sdk::setup_alloc!();
-
-const NO_DEPOSIT: Balance = 0;
 
 /// Gas to call finish withdraw method on factory.
-const FINISH_WITHDRAW_GAS: Gas = 50_000_000_000_000;
+const FINISH_WITHDRAW_GAS: Gas = Gas(50_000_000_000_000);
```
Comment (sept-en) [inline bridge-token/src/lib.rs:14]: Please use the following pattern for readability
```suggestion
const FINISH_WITHDRAW_GAS: Gas = Gas(Gas::ONE_TERA.0 * 50);
```

File: bridge-token-factory/src/lib.rs [hunk 3]
```diff
@@ -28,24 +27,24 @@ const BRIDGE_TOKEN_BINARY: &'static [u8] = include_bytes!(std::env!(
...
+const FINISH_UPDATE_METADATA_GAS: Gas = Gas(5_000_000_000_000);
 
 /// Amount of gas used by set_metadata in the factory, without taking into account
 /// the gas consumed by the promise.
-const OUTER_SET_METADATA_GAS: Gas = 15_000_000_000_000;
+const OUTER_SET_METADATA_GAS: Gas = Gas(15_000_000_000_000);
 
 /// Amount of gas used by bridge token to set the metadata.
-const SET_METADATA_GAS: Gas = 5_000_000_000_000;
+const SET_METADATA_GAS: Gas = Gas(5_000_000_000_000);
```
Comment (sept-en) [inline bridge-token-factory/src/lib.rs:47]: Please use `near-sdk`'s constants for `Gas::TGas` and `ONE_NEAR` (see my comment below)

File: bridge-token-factory/src/lib.rs [hunk 4]
```diff
@@ -343,32 +336,26 @@ impl BridgeTokenFactory {
...
-            ),
+            Some(message) => ext_bridge_token::ext(self.get_bridge_token_account_id(token.clone()))
+                .with_static_gas(MINT_GAS)
+                .with_attached_deposit(env::attached_deposit() - required_deposit)
+                .mint(env::current_account_id(), amount.into())
+                .then(
+                    ext_bridge_token::ext(self.get_bridge_token_account_id(token))
+                        .with_static_gas(FT_TRANSFER_CALL_GAS)
+                        .with_attached_deposit(1)
+                        .ft_transfer_call(target.try_into().unwrap(), amount.into(), None, message),
```
Comment (sept-en) [inline bridge-token-factory/src/lib.rs:353]: `target` is already `AccountId` type
```suggestion
                        .ft_transfer_call(target.unwrap(), amount.into(), None, message),
```

File: bridge-token-factory/src/lib.rs [hunk 5]
```diff
@@ -383,9 +370,9 @@ impl BridgeTokenFactory {
         #[serializer(borsh)] recipient: String,
     ) -> ResultType {
         let token = env::predecessor_account_id();
-        let parts: Vec<&str> = token.split(".").collect();
+        let parts: Vec<&str> = token.as_str().split(".").collect();
```
Comment (sept-en) [inline bridge-token-factory/src/lib.rs:373]:
```suggestion
        let parts: Vec<&str> = token.as_str().split('.').collect();
```

**Review by sept-en — CHANGES_REQUESTED** [review]

File: bridge-token/src/lib.rs (hunk 2 above)
Comment (karim-en) [reply bridge-token/src/lib.rs]: Fixed 8e2b316371f19907869b1bd8506f976beb82ce31

File: bridge-token-factory/src/lib.rs (hunk 3 above)
Comment (karim-en) [reply bridge-token-factory/src/lib.rs]: Fixed 8e2b316371f19907869b1bd8506f976beb82ce31

File: bridge-token-factory/src/lib.rs (hunk 4 above)
Comment (karim-en) [reply bridge-token-factory/src/lib.rs]: Fixed 8e2b316371f19907869b1bd8506f976beb82ce31

File: bridge-token-factory/src/lib.rs (hunk 5 above)
Comment (karim-en) [reply bridge-token-factory/src/lib.rs]: Fixed 8e2b316371f19907869b1bd8506f976beb82ce31

**Pull request merged**